Chapter_3_Keyword_Design_Variables_Resources.md

```diff
@@ -0,0 +1,1066 @@
+
+
+
+
+## 3.3 User Keyword Definition & Arguments
+
+User Keywords in Robot Framework allow users to create their own
+keywords by combining existing keywords into reusable higher-level actions.
+They help improve readability, maintainability, and modularity in
+automation by abstracting complex sequences into named actions.
+User Keywords are defined syntactically very similarly to tests|tasks
+and are defined in the `*** Keywords ***` section of a suite file or resource file.
+
+
+
+### 3.3.1 `*** Keywords ***` Section
+
+The `*** Keywords ***` section of suite and resource files
+is indentation-based similar to the `*** Test Cases ***` section.
+The user keywords defined are unindented, while their body implementation is indented by multiple spaces.
+
+See these sections for more details about
+[2.2 Basic Suite File Syntax](Chapter_2_Getting_Started.md#22-basic-suite-file-syntax)
+and [2.6 Writing Test|Task and Calling Keywords](Chapter_2_Getting_Started.md#26-writing-testtask-and-calling-keywords).
+
+This section can be part of suites or resource files.
+While keywords defined in suites can solely be used in the suite they are defined in,
+keywords defined in resource files can be used in any suite that imports these resource files.
+
+Example definition of a user keyword:
+
+```robotframework
+*** Keywords ***
+Verify Valid Login
+    [Arguments]    ${exp_full_name}
+    ${version}=    Get Server Version
+    Should Not Be Empty    ${version}
+    ${name}=    Get User Name
+    Should Be Equal    ${name}    ${exp_full_name}
+```
+
+As a reference for how defined keywords are documented, see [2.5 Keyword Interface and Documentation](Chapter_2_Getting_Started.md#25-keyword-interface-and-documentation).
+
+
+
+### 3.3.2 User Keyword Names
+
+> [!IMPORTANT]
+> LO-3.3.2 Recall the rules how keyword names are matched. (K1)
+
+The names of User Keywords should be descriptive and clear, reflecting the purpose of the keyword.
+Well-named keywords make tests more readable and easier to understand.
+Robot Framework supports Unicode and allows the use of special characters and even Emojis in keyword names.
+
+Keyword names are case-insensitive and can include spaces.
+Also spaces and underscores will be ignored when matching keyword names.
+So the keywords `Login To System`, and `log_into_system` are considered identical.
+
+To identify keywords that shall be executed, Robot Framework uses a matching algorithm that is case-insensitive and ignores spaces and underscores.
+If then a full match is found, that keyword is used.
+If no full match is found, the prefixes `Given`, `When`, `Then`, `And`, and `But` (case-insensitive), which are used in Behavior-Driven Specification style, are removed from the called keyword name to find a match.
+If still no match is found, Robot Framework tries to match the name with keywords that have embedded arguments.
```

Comment:
Orderded (numbered) list could work better. Rendered version with all rules collapsed into a single paragraph was more difficult to read.